pkg/cloud/services/ec2/launchtemplate_test.go

```diff
@@ -309,3 +436,850 @@ func TestService_LaunchTemplateNeedsUpdate(t *testing.T) {
+						UserData:         pointer.StringPtr(base64.StdEncoding.EncodeToString(userData)),
+						SecurityGroupIds: []*string{aws.String("1"), aws.String("2")},
+						ImageId:          aws.String("imageID"),
+						TagSpecifications: []*ec2.LaunchTemplateTagSpecificationRequest{
+							{
+								ResourceType: aws.String(ec2.ResourceTypeInstance),
+								Tags:         defaultEC2Tags("aws-mp-name", "cluster-name"),
+							},
+							{
+								ResourceType: aws.String(ec2.ResourceTypeVolume),
+								Tags:         defaultEC2Tags("aws-mp-name", "cluster-name"),
+							},
+						},
+					},
+					LaunchTemplateName: aws.String("aws-mp-name"),
+					TagSpecifications: []*ec2.TagSpecification{
+						{
+							ResourceType: aws.String(ec2.ResourceTypeLaunchTemplate),
+							Tags:         defaultEC2Tags("aws-mp-name", "cluster-name"),
+						},
+					},
+				}
+				m.CreateLaunchTemplate(gomock.AssignableToTypeOf(expectedInput)).Return(nil,
+					awserrors.NewFailedDependency("dependency failure")).Do(func(arg *ec2.CreateLaunchTemplateInput) {
+					// formatting added to match arrays during reflect.DeepEqual
+					formatTagsInput(arg)
+					if !reflect.DeepEqual(expectedInput, arg) {
+						t.Fatalf("mismatch in input expected: %+v, got: %+v", expectedInput, arg)
+					}
+				})
+			},
+			check: func(id string, err error) {
+				if id != "" {
+					t.Fatalf("expected empty launch template id, but got %+v", id)
+				}
+				if err == nil {
+					t.Fatalf("expected an error but got nothing")
+				}
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			scheme, err := setupScheme()
+			if err != nil {
+				t.Fatalf("failed during scheme registration")
+			}
+
+			client := fake.NewClientBuilder().WithScheme(scheme).Build()
+
+			mockScope := mocks.NewMockEC2Scope(mockCtrl)
+			mockEC2Client := mock_ec2iface.NewMockEC2API(mockCtrl)
+			mockSSMClient := mock_ssmiface.NewMockSSMAPI(mockCtrl)
+
+			mpScope, err := setupMachinePoolScope(client, mockScope)
+			if err != nil {
+				t.Fatalf("failed while creating machinepool scope with err: %+v", err)
+			}
+
+			s := &Service{
+				scope:     mockScope,
+				EC2Client: mockEC2Client,
+				SSMClient: mockSSMClient,
+			}
+
+			tc.expect(mockScope.EXPECT(), mockEC2Client.EXPECT())
+			tc.check(s.CreateLaunchTemplate(mpScope, aws.String("imageID"), userData))
+		})
+	}
+}
+
+func TestCreateLaunchTemplateVersion(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	var formatTagsInput = func(arg *ec2.CreateLaunchTemplateVersionInput) {
+		for index := range arg.LaunchTemplateData.TagSpecifications {
+			sortTags(arg.LaunchTemplateData.TagSpecifications[index].Tags)
+		}
+	}
+
+	var userData = []byte{1, 0, 0}
+	testCases := []struct {
+		name    string
+		imageID *string
+		expect  func(scope *mocks.MockEC2ScopeMockRecorder, m *mock_ec2iface.MockEC2APIMockRecorder)
+		check   func(e error)
+	}{
+		{
```

Comment:
Is it possible to add following scenario here:
Should fail to form launch template data

Reply:
done